src/coprocessor/time_limiter.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use pin_project::pin_project;
+use std::future::Future;
+use std::marker::PhantomData;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::{Duration, Instant};
+use tokio::sync::{Semaphore, SemaphorePermit};
+
+use super::metrics::*;
+
+/// Limits the time used by `fut` before a semaphore permit is acquired.
+///
+/// The future `fut` can always run for at least `time_limit_without_permit`,
+/// but it needs to acquire a permit from the semaphore before it can continue.
+pub fn limit_time<'a, F: Future + 'a>(
```

Comment:
I think the name `limit_time` is vague. It is not a mechanism to limit the execution time but a concurrent limiter (time limit is its method, but not the purpose). Maybe we can call it `ConcurrentGate` or `ConcurrentLimiter`.